arroyo/processing/strategies/run_task.py

```diff
@@ -0,0 +1,97 @@
+import logging
+import time
+from collections import deque
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Callable, Deque, Optional, Tuple, TypeVar
+
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import Commit, Message, TPayload
+
+logger = logging.getLogger(__name__)
+
+TOutput = TypeVar("TOutput")
+
+
+class RunTask(ProcessingStrategy[TPayload]):
+    """
+    This strategy can be used to run IO-bound tasks in parallel, then commit offsets.
+
+    The user specifies a processing function (a callable that takes a message). For each message received
+    in the submit method, it runs that processing function. Once completed, offsets are committed.
+    Note that the return value of the processing function is discarded so this is not an appropriate
+    strategy if something else needs to happen after offsets are committed and the function returns.
+
+    If there are too many pending futures, we MessageRejected will be raised to notify the stream processor
+    to slow down.
+
+    On poll we check for completion of futures. If processing is done, we commit the offsetes.
+    If an error occured the original exception will be raised.
+
+    Caution: MessageRejected is not properly handled by the ParallelTransform step. Exercise
+    caution if chaining this step anywhere after a parallel transform.
+    """
+
+    def __init__(
+        self,
+        processing_function: Callable[[Message[TPayload]], TOutput],
+        concurrency: int,
+        max_pending_futures: int,
+        commit: Commit,
+    ) -> None:
+        self.__executor = ThreadPoolExecutor(max_workers=concurrency)
+        self.__function = processing_function
+        self.__queue: Deque[Tuple[Message[TPayload], Future[TOutput]]] = deque()
+        self.__max_pending_futures = max_pending_futures
+        self.__commit = commit
+        self.__closed = False
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+        # The list of pending futures is too long, tell the stream processor to slow down
+        if len(self.__queue) > self.__max_pending_futures:
+            raise MessageRejected
+
+        self.__queue.append((message, self.__executor.submit(self.__function, message)))
+
+    def poll(self) -> None:
+        while self.__queue:
+            message, future = self.__queue.popleft()
+
+            if not future.done():
+                break
+
+            exc = future.exception()
+
+            if exc is not None:
+                raise exc
```

Comment:
nit: if you call the `result` method on a future and the future contains an exception the exception will be raised. 
No need to extract the exception and re-rise it manually. `result` does it for you.

Reply:
Updated